cores/esp8266/Stream.h

```diff
@@ -102,12 +113,104 @@ class Stream: public Print {
         virtual String readString();
         String readStringUntil(char terminator);
 
+        virtual int read (uint8_t* buffer, size_t len);
+        int read (char* buffer, size_t len) { return read((uint8_t*)buffer, len); }
+
+        //////////////////// extension: direct access to input buffer
+        // to provide when possible a pointer to available data for read
+
+        // informs user and ::to*() on effective buffered peek API implementation
+        // by default: not available
+        virtual bool peekBufferAPI () const { return false; }
+
+        // returns number of byte accessible by peekBuffer()
+        virtual size_t peekAvailable () { return 0; }
+
+        // returns a pointer to available data buffer (size = peekAvailable())
+        // semantic forbids any kind of ::read()
+        //     - after calling peekBuffer()
+        //     - and before calling peekConsume()
+        virtual const char* peekBuffer () { return nullptr; }
+
+        // consumes bytes after peekBuffer() use
+        // (then ::read() is allowed)
+        virtual void peekConsume (size_t consume) { (void)consume; }
+
+        // by default read timeout is possible (incoming data from network,serial..)
+        // children can override to false (like String::)
+        virtual bool inputTimeoutPossible () { return true; }
```

Comment:
> Why are these public?

OK to make them protected.

> I think that what you really want is to have a StreamPeekBuffer class in between that implements this api.

I don't think I want that 😆 
Given the way these classes are written, it is preferable to rely on overloading and keep the inheritance line straight. But I agree that in a template world, what you suggests would have been the preferred way.

Reply:
It is not a good idea to expose everything in a class. The private/protected keywords exist for very good reasons. You should hide implementation details whenever possible, especially in cases where methods are really internal details.
I'll ask in a different way: what would be the purpose of exposing these as public? What would be a sane use case in a sketch? I can think of use cases in objects that inherit from this, but it doesn't make sense to me to leave these open to the end user implementing a sketch.

>keep the inheritance line straight.

The line _is_ being kept straight, hierarchy is straight top to bottom. I am just suggesting to implement common functionality in a base class put in between. That is trivial to implement, and it expresses your real intent: 
- some streams are buffered, so they should inherit from StreamPeekBuffered, i.e.: Stream + peekbuffer
- some streams are not buffered, so they should inherit directly from Stream

The point here is that Streams that aren't buffered shouldn't be polluted with dummy implementations of these methods that don't apply to them.

3rd party libs that have objects that inherit from Stream won't be using these methods in their standard implementations, so they should inherit directly from Stream anyways. Unless they actually want to have implementations specific to our ESP and our peekbuffer, in which case they should inherit from out StreamPeekBuffer. From an user interface point of view, handling via a Stream base pointer is still ok for both case, as it should be.

Reply:
I don't get it, because `Client` derives from `Stream`, some `Client` wouldn't implements these extensions while others would.
How is an hypothetical `StreamDirect` class supposed to be kept in a straight line ?
I think that would add more confusion to the API.

The peekBuffer API is in the public interface to make it accessible by users. By these functions output buffers are directly accessible without copies. Why would we want to hide it ?